---
toc: true 
comments: true 
layout: post 
title: Algorithmic Code Prep 
type: hacks 
courses: { csse: {week: 1}, csp: {week: 1, categories: [4.A]}, csa: {week: 26}} 
---

<h2> Our Algorithmic Dance! </h2>

- Our whole gig is that there is this very sought out for bachelorette named Rachilita and there is a roster of 12 bachelors trying to woo her with code pick up lines
- Rachilita ratings each bachelors' pick up line from 1 to 12 (1 the best, 12 the worst) and needs our help to sort them all in order!!
- I was ratinged 6 and only had to move once --> from this now I will never forget how bubble sort works!!!

<img src="https://i.imgur.com/969lKO1.png" width="600">


<h2> Learn all sort </h2>

- Build into this Jupyter Notebook(s) for Bubble Sort, Selection Sort, Insertion Sort, Merge Sort, Quick Sort. Share solutions.
- Build into this Jupyter Notebooks Tester methods that runs each Sort
- Build your sorts into Generic T List (aka Collectable) extending Comparable using compareTo to sort the objects.

<mark>Notes</mark>: 

- <b>implementing Comparable interface</b>: a class gains the ability to compare its instances with other instances of the same class --> useful for sorting and ordering objects, as it provides a standard way to determine their relative positions

- <b>compareTo method</b>: compares 2 objects of the same class and determines their relative ordering 

1. if the current object is less than the object being compared to, compareTo should return a negative integer.
2. if the current object is equal to the object being compared to, compareTo should return 0.
3. ithe current object is greater than the object being compared to, compareTo should return a positive integer.

<h3>Comparable object for our Bachelors </h3>

In [1]:
import java.util.ArrayList;
import java.util.Collections;

public class Bachelor implements Comparable<Bachelor> {
    private String name;
    private int rating;

    // overloaded constructor 
    public Bachelor(String name, int rating) {
        this.name = name;
        this.rating = rating;
    }

    // getter methods
    public String getName() {
        return name;
    }

    public int getRating() {
        return rating;
    }

    @Override
    // compareTo method
    public int compareTo(Bachelor other) {
        return Integer.compare(this.rating, other.rating);
    }

    // list of predefined names for Bachelors
    private static final List<String> NAMES = List.of(
        "Justin", "Finn", "Theo", "Grace", "Emma", "Luna",
        "Aliya", "Vivian", "Shivansh", "Pranavi", "Rayhan", "Isabelle",
        "Natalie", "Oscar", "Penelope",
        "Quentin", "Rebecca", "Sophia", "Tyler", "Ursula", "Violet",
        "William", "Xander", "Yvonne", "Zara");

    // method to create a list of random Bachelor objects
    public static ArrayList<Bachelor> createList() {
        ArrayList<Bachelor> bachelorList = new ArrayList<>();
        for (int i = 0; i < 10000; i++) {
            String name = NAMES.get(i % NAMES.size()); // get a name from the predefined list
            int rating = (int) (Math.random() * 10000) + 1; // generate a random rating between 1 and 10,000
            bachelorList.add(new Bachelor(name, rating)); // create and add a new Bachelor object to the list
        }
        return bachelorList;
    }
}


<h3> Bubble Sort 🫧 </h3>

I feel like the name of this algorithm makes the logic of this sort very easy to comprehend --> it starts from the first element of an array and compares it with the second element. If the first element is greater than the second, we swap them. It continues this process until the end of the array, with the largest elements “bubbling” to the top 


In [2]:
import java.util.ArrayList;
import java.util.Collections;

public class BubbleSort {
    public static void bubbleSort(ArrayList<Bachelor> bachelors) {
        int n = bachelors.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (bachelors.get(j).compareTo(bachelors.get(j + 1)) > 0) {
                    // swap bachelors
                    Collections.swap(bachelors, j, j + 1);
                } 
            }
        }
    }

    public static void main(String[] args) {
        // create a list of Bachelor objects
        ArrayList<Bachelor> bachelorList = Bachelor.createList();

        // measure sorting time
        long startTime = System.nanoTime();

        // sort the list using Bubble Sort
        bubbleSort(bachelorList);

        long endTime = System.nanoTime();
        long duration = (endTime - startTime);  // Elapsed time in nanoseconds

        System.out.println("\nTime taken to sort: " + duration + " nanoseconds\n");

        // print sorted list
        System.out.println("\nSorted List:");
        for (Bachelor bachelor : bachelorList) {
            System.out.println("Bachelor: " + bachelor.getName() + "| Rating: " + bachelor.getRating());
        }
    }
}

BubbleSort.main(null);


Time taken to sort: 646910042 nanoseconds


Sorted List:
Bachelor: Finn| Rating: 3
Bachelor: Finn| Rating: 3
Bachelor: Yvonne| Rating: 4
Bachelor: Sophia| Rating: 7
Bachelor: Rebecca| Rating: 7
Bachelor: Shivansh| Rating: 10
Bachelor: Quentin| Rating: 12
Bachelor: Penelope| Rating: 12
Bachelor: Ursula| Rating: 12
Bachelor: Luna| Rating: 14
Bachelor: Theo| Rating: 14
Bachelor: Ursula| Rating: 14
Bachelor: Grace| Rating: 15
Bachelor: Quentin| Rating: 17
Bachelor: Pranavi| Rating: 20
Bachelor: William| Rating: 21
Bachelor: Justin| Rating: 23
Bachelor: Violet| Rating: 24
Bachelor: Natalie| Rating: 25
Bachelor: Natalie| Rating: 26
Bachelor: Penelope| Rating: 26
Bachelor: Vivian| Rating: 27
Bachelor: Ursula| Rating: 27
Bachelor: Isabelle| Rating: 27
Bachelor: Pranavi| Rating: 29
Bachelor: Finn| Rating: 30
Bachelor: Natalie| Rating: 30
Bachelor: Violet| Rating: 31
Bachelor: Penelope| Rating: 32
Bachelor: Ursula| Rating: 33
Bachelor: Quentin| Rating: 33
Bachelor: Rebecca| Rating: 34
Bachelor:

<h3>Insertion Sort 🂠 </h3>

Works by iteratively inserting each element of an unsorted list into its correct position in a sorted portion of the list. It's kind of like sorting playing cards in your hands. You split the cards into two groups: the sorted cards and the unsorted cards. Then, you pick a card from the unsorted group and put it in the right place in the sorted group.

In [7]:
import java.util.ArrayList;

public class InsertionSort {
    public static void insertionSort(ArrayList<Bachelor> bachelors) {
        int n = bachelors.size();
        for (int i = 1; i < n; i++) {
            Bachelor key = bachelors.get(i);
            int j = i - 1;
            while (j >= 0 && bachelors.get(j).compareTo(key) > 0) {
                bachelors.set(j + 1, bachelors.get(j));
                j = j - 1;
            }
            bachelors.set(j + 1, key);
        }
    }

    public static void main(String[] args) {
        // create a list of Bachelor objects
        ArrayList<Bachelor> bachelorList = Bachelor.createList();

        // measure sorting time
        long startTime = System.nanoTime();

        // sort the list using Insertion Sort
        insertionSort(bachelorList);

        long endTime = System.nanoTime();
        long duration = (endTime - startTime);  // Elapsed time in nanoseconds

        System.out.println("\nTime taken to sort: " + duration + " nanoseconds");
        
        // print sorted list
        System.out.println("\nSorted List:");
        for (Bachelor bachelor : bachelorList) {
            System.out.println("Bachelor: " + bachelor.getName() + "| Rating: " + bachelor.getRating());
        }
    }
}
InsertionSort.main(null);


Time taken to sort: 156783500 nanoseconds

Sorted List:
Bachelor: Finn| Rating: 1
Bachelor: Natalie| Rating: 2
Bachelor: Yvonne| Rating: 3
Bachelor: Grace| Rating: 4
Bachelor: Aliya| Rating: 4
Bachelor: Shivansh| Rating: 6
Bachelor: Luna| Rating: 6
Bachelor: Theo| Rating: 7
Bachelor: Quentin| Rating: 7
Bachelor: William| Rating: 8
Bachelor: Ursula| Rating: 8
Bachelor: Tyler| Rating: 9
Bachelor: Aliya| Rating: 9
Bachelor: Rebecca| Rating: 9
Bachelor: Justin| Rating: 11
Bachelor: Violet| Rating: 11
Bachelor: Vivian| Rating: 12
Bachelor: Shivansh| Rating: 16
Bachelor: Xander| Rating: 17
Bachelor: Vivian| Rating: 18
Bachelor: Theo| Rating: 19
Bachelor: Penelope| Rating: 19
Bachelor: Grace| Rating: 19
Bachelor: Quentin| Rating: 20
Bachelor: Violet| Rating: 20
Bachelor: Emma| Rating: 21
Bachelor: Penelope| Rating: 21
Bachelor: Aliya| Rating: 25
Bachelor: Luna| Rating: 28
Bachelor: Aliya| Rating: 29
Bachelor: Xander| Rating: 31
Bachelor: Natalie| Rating: 31
Bachelor: Oscar| Rating: 32
Bachel

<h3>Selection Sort ⭐ </h3>

The first smallest element is selected from the unsorted array and placed at the first position. After that second smallest element is selected and placed in the second position. The process continues until the array is entirely sorted.


In [8]:
import java.util.ArrayList;

public class SelectionSort {
    public static void selectionSort(ArrayList<Bachelor> bachelors) {
        int n = bachelors.size();
        for (int i = 0; i < n - 1; i++) {
            int minIndex = i;
            for (int j = i + 1; j < n; j++) {
                if (bachelors.get(j).compareTo(bachelors.get(minIndex)) < 0) {
                    minIndex = j;
                }
            }
            // swap the minimum element with the first element of the unsorted subarray
            Bachelor temp = bachelors.get(minIndex);
            bachelors.set(minIndex, bachelors.get(i));
            bachelors.set(i, temp);
        }
    }

    public static void main(String[] args) {
        // create a list of Bachelor objects
        ArrayList<Bachelor> bachelorList = Bachelor.createList();

        // measure sorting time
        long startTime = System.nanoTime();
        
        // sort the list using Selection Sort
        selectionSort(bachelorList);

        long endTime = System.nanoTime();
        long duration = (endTime - startTime);  // Elapsed time in nanoseconds

        System.out.println("\nTime taken to sort: " + duration + " nanoseconds");

        // print sorted list
        System.out.println("\nSorted List:");
        for (Bachelor bachelor : bachelorList) {
            System.out.println("Bachelor: " + bachelor.getName() + "| Rating: " + bachelor.getRating());
        }
    }
}
SelectionSort.main(null);


Time taken to sort: 127719416 nanoseconds

Sorted List:
Bachelor: Xander| Rating: 1
Bachelor: Shivansh| Rating: 1
Bachelor: William| Rating: 5
Bachelor: Grace| Rating: 8
Bachelor: Sophia| Rating: 8
Bachelor: Xander| Rating: 8
Bachelor: Theo| Rating: 8
Bachelor: Sophia| Rating: 8
Bachelor: Grace| Rating: 9
Bachelor: Emma| Rating: 9
Bachelor: Penelope| Rating: 10
Bachelor: Rayhan| Rating: 11
Bachelor: Yvonne| Rating: 12
Bachelor: Rayhan| Rating: 13
Bachelor: Vivian| Rating: 14
Bachelor: Luna| Rating: 17
Bachelor: Pranavi| Rating: 17
Bachelor: Xander| Rating: 17
Bachelor: Quentin| Rating: 18
Bachelor: Sophia| Rating: 18
Bachelor: Aliya| Rating: 19
Bachelor: Quentin| Rating: 19
Bachelor: Quentin| Rating: 20
Bachelor: Grace| Rating: 21
Bachelor: Sophia| Rating: 22
Bachelor: Luna| Rating: 24
Bachelor: Violet| Rating: 26
Bachelor: Natalie| Rating: 26
Bachelor: Shivansh| Rating: 26
Bachelor: Finn| Rating: 27
Bachelor: William| Rating: 27
Bachelor: Yvonne| Rating: 29
Bachelor: Finn| Rating: 30

<h3> Merge Sort 🤜 🤛</h3>

Follows the divide-and-conquer approach. It works by recursively dividing the input array into smaller subarrays and sorting those subarrays then merging them back together to obtain the sorted array. This one sounds like the name too tbh

In [5]:
import java.util.ArrayList;

public class MergeSort {

    public static void mergeSort(ArrayList<Bachelor> bachelors) {
        if (bachelors.size() > 1) {
            int mid = bachelors.size() / 2;

            // split the list into two halves
            ArrayList<Bachelor> leftHalf = new ArrayList<>(bachelors.subList(0, mid));
            ArrayList<Bachelor> rightHalf = new ArrayList<>(bachelors.subList(mid, bachelors.size()));

            // recursively sort each half
            mergeSort(leftHalf);
            mergeSort(rightHalf);

            // merge the sorted halves
            int i = 0, j = 0, k = 0;
            while (i < leftHalf.size() && j < rightHalf.size()) {
                if (leftHalf.get(i).compareTo(rightHalf.get(j)) <= 0) {
                    bachelors.set(k++, leftHalf.get(i++));
                } else {
                    bachelors.set(k++, rightHalf.get(j++));
                }
            }

            // copy remaining elements from left half
            while (i < leftHalf.size()) {
                bachelors.set(k++, leftHalf.get(i++));
            }

            // copy remaining elements from right half
            while (j < rightHalf.size()) {
                bachelors.set(k++, rightHalf.get(j++));
            }
        }
    }

    public static void main(String[] args) {
        ArrayList<Bachelor> bachelorList = Bachelor.createList();
        
        // measure sorting time
        long startTime = System.nanoTime();

        // applying merge sort
        mergeSort(bachelorList);

        long endTime = System.nanoTime();
        long duration = (endTime - startTime);  // Elapsed time in nanoseconds

        System.out.println("\nTime taken to sort: " + duration + " nanoseconds");

        // printing the sorted list of bachelors
        System.out.println("\nSorted List:");
        for (Bachelor bachelor : bachelorList) {
            System.out.println(bachelor.getName() + " (" + bachelor.getRating() + ")");
        }
    }
}

MergeSort.main(null);


Time taken to sort: 13339875 nanoseconds

Sorted List:
Emma (3)
Theo (3)
Ursula (4)
Rayhan (5)
Oscar (5)
Natalie (6)
Yvonne (7)
Ursula (7)
Rayhan (7)
Justin (7)
Luna (8)
Isabelle (10)
Zara (10)
Quentin (10)
Zara (10)
Isabelle (12)
Xander (14)
Penelope (15)
Theo (15)
Vivian (18)
Violet (19)
Yvonne (20)
Xander (20)
Finn (20)
Finn (21)
Quentin (21)
Aliya (25)
Oscar (25)
Zara (26)
Rayhan (26)
Penelope (27)
Emma (28)
Emma (29)
Quentin (30)
Emma (31)
Ursula (32)
Theo (33)
Tyler (34)
Finn (34)
Oscar (35)
Isabelle (35)
Theo (35)
Oscar (36)
Pranavi (36)
Rebecca (37)
William (40)
Vivian (41)
Ursula (42)
Penelope (42)
Vivian (45)
Vivian (48)
Justin (53)
Tyler (54)
Natalie (55)
Violet (58)
Grace (59)
Isabelle (62)
Oscar (64)
Shivansh (64)
Violet (64)
Rebecca (65)
Violet (66)
Rayhan (67)
Emma (68)
Penelope (70)
Ursula (71)
Emma (71)
Aliya (72)
Oscar (72)
Vivian (74)
Shivansh (74)
Grace (74)
Xander (76)
Quentin (79)
Oscar (79)
William (79)
Rayhan (81)
Justin (83)
Vivian (83)
Yvonne (84)
Violet (84)

<h3> Quick Sort 🌬️</h3>

Selects a pivot element and then sorts values larger than it on one side and smaller to the other side, and then it repeats those steps until the array is sorted. Useful for sorting big data sets

In [6]:
import java.util.ArrayList;

public class QuickSort {

    public static void quickSort(ArrayList<Bachelor> bachelors, int low, int high) {
        if (low < high) {
            // Partition the array
            int pi = partition(bachelors, low, high);

            // Recursively sort elements before partition and after partition
            quickSort(bachelors, low, pi - 1);
            quickSort(bachelors, pi + 1, high);
        }
    }

    private static int partition(ArrayList<Bachelor> bachelors, int low, int high) {
        Bachelor pivot = bachelors.get(high);
        int i = low - 1; // index of smaller element

        for (int j = low; j < high; j++) {
            // If current element is smaller than the pivot
            if (bachelors.get(j).compareTo(pivot) < 0) {
                i++;

                // Swap bachelors[i] and bachelors[j]
                Bachelor temp = bachelors.get(i);
                bachelors.set(i, bachelors.get(j));
                bachelors.set(j, temp);
            }
        }

        // Swap bachelors[i+1] and pivot
        Bachelor temp = bachelors.get(i + 1);
        bachelors.set(i + 1, bachelors.get(high));
        bachelors.set(high, temp);

        return i + 1;
    }

    public static void main(String[] args) {
        ArrayList<Bachelor> bachelorList = Bachelor.createList();
        
        // measure sorting time
        long startTime = System.nanoTime();
        
        // Applying quick sort
        quickSort(bachelorList, 0, bachelorList.size() - 1);

        long endTime = System.nanoTime();
        long duration = (endTime - startTime);  // Elapsed time in nanoseconds

        System.out.println("\nTime taken to sort: " + duration + " nanoseconds");

        // Printing the sorted list of bachelors
        System.out.println("\nSorted List:");
        for (Bachelor bachelor : bachelorList) {
            System.out.println(bachelor.getName() + " (" + bachelor.getRating() + ")");
        }
    }
}

QuickSort.main(null);


Time taken to sort: 6719875 nanoseconds

Sorted List:
Emma (1)
Natalie (1)
Penelope (1)
Theo (2)
Penelope (2)
Finn (2)
Pranavi (4)
Violet (5)
Vivian (6)
Theo (6)
Yvonne (7)
Isabelle (8)
Emma (8)
Violet (8)
Yvonne (9)
Shivansh (10)
Natalie (13)
Quentin (16)
Vivian (17)
Penelope (18)
Theo (20)
Justin (20)
Rayhan (21)
Emma (22)
Pranavi (22)
Pranavi (23)
Sophia (23)
Theo (25)
Zara (27)
Grace (28)
Isabelle (28)
Natalie (28)
Zara (30)
William (30)
Emma (30)
William (31)
Yvonne (31)
Finn (31)
Grace (32)
Rebecca (33)
Tyler (35)
Sophia (36)
Grace (39)
Tyler (39)
Penelope (40)
Quentin (40)
Quentin (40)
Xander (41)
Luna (44)
Zara (45)
Xander (45)
Emma (45)
Penelope (46)
Vivian (46)
Rebecca (46)
Oscar (46)
Theo (47)
Pranavi (47)
Penelope (48)
Pranavi (49)
Justin (49)
Zara (50)
Sophia (52)
Zara (54)
Pranavi (55)
Sophia (56)
Finn (56)
Emma (57)
William (57)
Pranavi (58)
Rayhan (59)
Violet (63)
Penelope (63)
Aliya (64)
Natalie (64)
Zara (65)
Oscar (66)
Finn (66)
Luna (68)
Luna (71)
Rayhan (71)
Tyler

<h3> Notes on Sorts' efficiency </h3>

1. Quick Sort: ~6,719,875 👑👑👑👑👑👑👑
- Average Time Complexity: O(n log n)
- Quick sort typically performs very well in practice and is often the preferred choice for sorting large datasets due to its average-case time complexity of O(n log n).

2. Merge Sort: ~13,339,875
- Average Time Complexity: O(n log n)
- Merge sort also has an average-case time complexity of O(n log n) and is known for its stability and predictable performance across different datasets.

3. Selection Sort: ~127,719,416
- Average Time Complexity: O(n^2)
- Works better than insertion when array is highly unsorted 

4. Insertion Sort: ~153,107,500    
- Average Time Complexity: O(n^2)
- Insertion sort performs reasonably well for small datasets and nearly sorted arrays. However, its average-case time complexity of O(n^2) makes it less efficient compared to quick sort and merge sort for large datasets.

5. Bubble Sort: ~646,910,042
- Average Time Complexity: O(n^2)
- Bubble sort is generally the least efficient of the five algorithms listed here, with a time complexity of O(n^2). It is mainly used for educational purposes or for sorting very small datasets due to its simplicity.